### ISSUE TITLE
Epics show up as "unestimated issues" in burndown chart

### ISSUE THREAD

**Issue by cbeams** [ZenHubIO/support#674]
Repro:

 - Create an epic. Do not assign it a story point value
 - Add two issues to the epic, each having their own story point value, say, of 1 and 2
 - Note that at this point, the epic is worth 3 story points—the sum of the two issues within it. This is not reflected in the epic's estimate field, but is reflected in the "0 of 3 story points" section.
 - Assign all three issues to the same milestone
 - Click 'Burndown'
 - Notice that there is "1 unestimated issue", referring to the Epic.

Expected behavior:
 - The epic would be understood to be (indirectly) estimated as a function of the issues within that epic all having been estimated.

Actual behavior:
 - The epic issue shows up as "unestimated" in a burndown context, making it seem as if something is incomplete, when in fact it is not.

I would argue that epics should not subject to direct estimation at all, i.e. should always derive their story point value from the issues within that epic, but perhaps direct estimation of epics is valid in some workflows. In any case, it would be nice if the burndown UI didn't make things appear as if there is something wrong for those who choose not to directly estimate epics.

**Comment by aaronwest** [comment]
I agree this is a topic worth discussion. From a Burndown chart perspective I too would expect the Epic issue to derive it's story points from each underlying issue.

A similar issue exists with the ZenHub API. I'm using the GitHub API + the ZenHub API to create all sorts of issue reports in a tabular format. Think project management type stuff. Before Epic issues my team would create a "master" issue and then add sub-issues via links in the master issue description. Instead of estimating each sub-issue we estimated the master issue as an aggregate. This was done so our reports (using the various APIs) could look at the master issue only to determine the value of the work completed.

Now, with Epics, it makes more sense to leave the Epic issue w/o an estimate in favor of estimating the sub-issues. This makes everything look nice and tidy on GitHub (with the exception of the Burndown charts as mentioned in this issue). However, it breaks our reporting as the behavior of the ZenHub API hasn't changed with the release of Epics.

This is probably a separate issue for the ZH team to address but since it's tangential to this issue I wanted to add this comment. My preference for the ZH API would be for it to report the aggregate estimate of issues inside an Epic issue. Similar to how @cbeams is requesting the Burndown charts behave.

**Comment by rajsesh-msft** [comment]
Currently you can add an issue to multiple epics.  If you are summing up the estimates of all issues in an epic, it would also make sense to enforce 1:N relationship between Epics and their children.  Adding an issue to a new epic should remove it from any other epic it is currently part of.

**Comment by cbeams** [comment]
> Adding an issue to a new epic should remove it from any other epic it is currently part of.

I wouldn't want this to be a hard rule. My team has already begun taking advantage of the fact that issues can be members of multiple epics. It's helpful in certain workflows, and I'm glad ZenHub allows it.

I agree it's an issue with regard to aggregating story point values at the epic level, though.

**Comment by Anvil102** [comment]
+1 to the idea that epics _should not_ be included in the main issue list of milestone's burndown as "_unestimated issues_."

It would be more useful to have epics broken out in separate column or box on the burndown report page as "Epics in the Milestone" or "Epics in the Sprint." Knowing how many epics a team closed (and the sum of their children's stroy points – and thus the scope of their size) in a sprint is valuable info for many teams to know @azenMatt

**Comment by azenMatt** [comment]
before we do this would it be useful to implement something so that when adding an epic to a milestone, ZenHub should ask if you would like to also add all the children of that epic?

----------

In terms of the story points totalling for the Epic, I think it makes sense to allow two numbers here. One that is user set at the Epic level. And one that is the sum of the children. 

We already have the sum of the children here:

![image](https://cloud.githubusercontent.com/assets/7584089/15511586/0a2a6e22-2191-11e6-9d85-f32a75d9b7c4.png)

But if it is useful to see on the board we can find a way to add it to the issue card 

I feel it is useful to keep allowing an epic to have it's own estimation as a way to quickly compare the size of epics that have not yet been fully fleshed out. Without making them look smaller than they actually are due to not having all child issues estimated yet. 

> I wouldn't want this to be a hard rule. My team has already begun taking advantage of the fact that issues can be members of multiple epics. It's helpful in certain workflows, and I'm glad ZenHub allows it.

I agree with this - and from a workflow perspective it makes sense that a single story would have impact on multiple epics. This can really help with prioritization if you know that completing one issue will have a positive impact on two (or more) areas.

**Comment by Anvil102** [comment]
@azenMatt 

>In terms of the story points totalling for the Epic, I think it makes sense to allow two numbers here. One that is user set at the Epic level. And one that is the sum of the children.

Interesting but I'm not sure I would actually use that feature much. I would say go for it if its easy to implement, but I would not proactively "ask" via a pop-up (that would get annoying fast). A simple button would suffice. When I click the button in the epic, all _open_ issues in the epic would be moved to the sprint the epic was in.

----

> In terms of the story points totalling for the Epic, I think it makes sense to allow two numbers here. One that is user set at the Epic level. And one that is the sum of the children.

Agreed. The current numbers ZH provides for epics (x of Y issues completed, A of B story points completed) are the valuable core metrics.

----

> But if it is useful to see on the board we can find a way to add it to the issue card

Maybe in gray text next to the "filter by epic" button? You definitely would not want them near the story point counter for traditional story points assigned directly to the epic, itself!

----

> I feel it is useful to keep allowing an epic to have it's own estimation as a way to quickly compare the size of epics that have not yet been fully fleshed out. Without making them look smaller than they actually are due to not having all child issues estimated yet.

Agreed. I like that epics can still have story points assigned to them directly. It provides maximum flexibility to allow teams to work within a context that makes sense for them. In some cases an agile team at Leap won't want to make an extra QA issue for an epic so they will put the QA estimate on the epic itself. That way, once all the children issues wrap up, the epic is moved into "ready for QA" and QA vets the whole epic at once.

----

> I wouldn't want this to be a hard rule. My team has already begun taking advantage of the fact that issues can be members of multiple epics. It's helpful in certain workflows, and I'm glad ZenHub allows it.
> I agree with this - and from a workflow perspective it makes sense that a single story would have impact on multiple epics. This can really help with prioritization if you know that completing one issue will have a positive impact on two (or more) areas.

+1 all the way.

----

**Comment by nickdruery** [comment]
soooo is it now possible to use EPICS with more granular issues assigned and complete a burndown chart on progress that works?

**Comment by aaronwest** [comment]
@azenMatt Matt, any thoughts on the timing of this issue? I'm still struggling to get any value from the Burndown charts since epic points aren't included (but I really want to be able to use them).

**Comment by simonmcl** [comment]
Need this feature too on ZenHub enterprise. We are using milestones for sprints (2 weeks) and for longer duration milestones like release cycles (~6  months). Epic points on the burndown graph would be perfect for us to automatically track our release cycle progress as well as sprint progress. Something management are asking for more visibility on.

Currently it doesn't well, if someone forgets to update the epic estimate, we get false positives in our reports. Also its very confusing as to why the epic points = (epic estimate + sum(child estimates)). Simply by assigning an estimate to the epic, to make it visible in the burndown, makes the epic twice as complex???

**Comment by aaronwest** [comment]
> Currently it doesn't well, if someone forgets to update the epic estimate, we get false positives in our reports. Also its very confusing as to why the epic points = (epic estimate + sum(child estimates)). Simply by assigning an estimate to the epic, to make it visible in the burndown, makes the epic twice as complex???

Yea, that's a large downside to Burndown charts right now that make them unusable for us. We are estimating issues at each issue level vs at the Epic issue level. This is the "correct" way to add estimates but Burndown charts don't take this into consideration.

When you look at individual ZenHub features such as Epics, story points/estimations, or the various charts they are pretty amazing. It's the integration between the individual features that leave me wanting and needing more.
